databroker/experimental/server_ext.py

```diff
@@ -0,0 +1,499 @@
+    register_query = query_registry.register
+
+    def __init__(
+        self,
+        *,
+        database,
+        directory,
+        queries=None,
+        sorting=None,
+        metadata=None,
+        principal=None,
+        access_policy=None,
+    ):
+        self.database = database
+        self.collection = database["reconstructions"]
+        self.directory = Path(directory).resolve()
+        if not self.directory.exists():
+            raise ValueError(f"Directory {self.directory} does not exist.")
+        if not self.directory.is_dir():
+            raise ValueError(
+                f"The given directory path {self.directory} is not a directory."
+            )
+        if not os.access(self.directory, os.W_OK):
+            raise ValueError("Directory {self.directory} is not writeable.")
+        self.queries = queries or []
+        self.sorting = sorting or [("metadata.scan_id", 1)]
+        self.metadata = metadata or {}
+        self.principal = principal
+        self.access_policy = access_policy
+        super().__init__()
+
+    @classmethod
+    def from_uri(cls, uri, directory, *, metadata=None):
+        if not pymongo.uri_parser.parse_uri(uri)["database"]:
+            raise ValueError(
+                f"Invalid URI: {uri!r} " f"Did you forget to include a database?"
+            )
+        client = pymongo.MongoClient(uri)
+        database = client.get_database()
+        return cls(database=database, directory=directory, metadata=metadata)
+
+    @classmethod
+    def from_mongomock(cls, directory, *, metadata=None):
+        import mongomock
+
+        db_name = f"temp-{str(uuid.uuid4())}"
+        mongo_client = mongomock.MongoClient()
+        database = mongo_client[db_name]
+
+        return cls(database=database, directory=directory, metadata=metadata)
+
+    def new_variation(
+        self,
+        metadata=UNCHANGED,
+        queries=UNCHANGED,
+        sorting=UNCHANGED,
+        principal=UNCHANGED,
+        **kwargs,
+    ):
+        if metadata is UNCHANGED:
+            metadata = self.metadata
+        if queries is UNCHANGED:
+            queries = self.queries
+        if sorting is UNCHANGED:
+            sorting = self.sorting
+        if principal is UNCHANGED:
+            principal = self.principal
+        return type(self)(
+            database=self.database,
+            directory=self.directory,
+            metadata=metadata,
+            queries=queries,
+            sorting=sorting,
+            access_policy=self.access_policy,
+            principal=principal,
+            **kwargs,
+        )
+
+    def post_metadata(self, metadata, structure_family, structure, specs, mimetype):
+        uid = str(uuid.uuid4())
+
+        validated_document = Document(
+            uid=uid,
+            structure_family=structure_family,
+            structure=structure,
+            metadata=metadata,
+            specs=specs,
+            mimetype=mimetype,
```

Comment:
`mimetype` should be defined above based on the structure. We always save array as `application/octet-stream` and we always save dataframe as `APACHE_ARROW_FILE_MIME_TYPE`.